File: lib/core/models/folder_profile.dart
import 'dart:convert';

class FolderProfile {
  final String id;
  String name;
  String folderPath;
  List<String> blacklistFolders;
  List<String> blacklistFiles;
  bool enabled;
  final DateTime createdAt;
  DateTime updatedAt;

  FolderProfile({
    required this.id,
    required this.name,
    required this.folderPath,
    List<String>? blacklistFolders,
    List<String>? blacklistFiles,
    this.enabled = true,
    DateTime? createdAt,
    DateTime? updatedAt,
  })  : blacklistFolders = blacklistFolders ?? [],
        blacklistFiles = blacklistFiles ?? [],
        createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  int get totalRules => blacklistFolders.length + blacklistFiles.length;

  FolderProfile copyWith({
    String? name,
    String? folderPath,
    List<String>? blacklistFolders,
    List<String>? blacklistFiles,
    bool? enabled,
  }) {
    return FolderProfile(
      id: id,
      name: name ?? this.name,
      folderPath: folderPath ?? this.folderPath,
      blacklistFolders: blacklistFolders ?? List.from(this.blacklistFolders),
      blacklistFiles: blacklistFiles ?? List.from(this.blacklistFiles),
      enabled: enabled ?? this.enabled,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'folderPath': folderPath,
        'blacklistFolders': blacklistFolders,
        'blacklistFiles': blacklistFiles,
        'enabled': enabled,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  factory FolderProfile.fromJson(Map<String, dynamic> json) => FolderProfile(
        id: json['id'] as String,
        name: json['name'] as String,
        folderPath: json['folderPath'] as String,
        blacklistFolders: List<String>.from(json['blacklistFolders'] ?? []),
        blacklistFiles: List<String>.from(json['blacklistFiles'] ?? []),
        enabled: json['enabled'] as bool? ?? true,
        createdAt: DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime.now(),
        updatedAt: DateTime.tryParse(json['updatedAt'] ?? '') ?? DateTime.now(),
      );

  String toJsonString() => jsonEncode(toJson());

  @override
  String toString() => 'FolderProfile($id, $name, $folderPath)';
}

File: SmartCopy-Rust/src-tauri/src/models/folder_profile.rs
use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FolderProfile {
    pub id: String,
    pub name: String,
    pub folder_path: String,
    pub blacklist_folders: Vec<String>,
    pub blacklist_files: Vec<String>,
    pub enabled: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl FolderProfile {
    pub fn new(id: String, name: String, folder_path: String) -> Self {
        let now = Utc::now();
        Self {
            id,
            name,
            folder_path,
            blacklist_folders: Vec::new(),
            blacklist_files: Vec::new(),
            enabled: true,
            created_at: now,
            updated_at: now,
        }
    }

    pub fn total_rules(&self) -> usize {
        self.blacklist_folders.len() + self.blacklist_files.len()
    }

    pub fn add_folder_rule(&mut self, pattern: String) {
        if !self.blacklist_folders.contains(&pattern) {
            self.blacklist_folders.push(pattern);
            self.updated_at = Utc::now();
        }
    }

    pub fn add_file_rule(&mut self, pattern: String) {
        if !self.blacklist_files.contains(&pattern) {
            self.blacklist_files.push(pattern);
            self.updated_at = Utc::now();
        }
    }

    pub fn remove_folder_rule(&mut self, pattern: &str) {
        self.blacklist_folders.retain(|p| p != pattern);
        self.updated_at = Utc::now();
    }

    pub fn remove_file_rule(&mut self, pattern: &str) {
        self.blacklist_files.retain(|p| p != pattern);
        self.updated_at = Utc::now();
    }

    pub fn with_folder_rules(mut self, patterns: Vec<String>) -> Self {
        self.blacklist_folders = patterns;
        self
    }

    pub fn with_file_rules(mut self, patterns: Vec<String>) -> Self {
        self.blacklist_files = patterns;
        self
    }

    pub fn matches_path(&self, path: &str) -> bool {
        let norm_src = path.replace('\\', "/").to_lowercase();
        let norm_profile_path = self.folder_path.replace('\\', "/").to_lowercase();
        norm_src.starts_with(&norm_profile_path)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProfilesData {
    pub profiles: Vec<FolderProfile>,
}

impl Default for ProfilesData {
    fn default() -> Self {
        Self {
            profiles: Vec::new(),
        }
    }
}

pub fn find_best_profile<'a>(profiles: &'a [FolderProfile], source_path: &str) -> Option<&'a FolderProfile> {
    let norm_src = source_path.replace('\\', "/").to_lowercase();
    let mut best: Option<&FolderProfile> = None;
    let mut best_depth: i32 = -1;

    for profile in profiles {
        if !profile.enabled {
            continue;
        }
        let norm_path = profile.folder_path.replace('\\', "/").to_lowercase();
        if norm_src.starts_with(&norm_path) {
            let depth = norm_path.split('/').count() as i32;
            if depth > best_depth {
                best_depth = depth;
                best = Some(profile);
            }
        }
    }

    best
}
